### PR TITLE
feat: switch to native Promise

### PR DESCRIPTION
**Pull request by alexkvak**

### CODE REVIEW

**Comment by kibertoad** [comment]
@alexkvak This needs to be synced with master after old Node versions were dropped

File: lib/channel_model.js [hunk 1]
```diff
@@ -74,10 +98,16 @@ C.open = function() {
 
 C.close = function() {
   var self = this;
-  return Promise.fromCallback(function(cb) {
+
+  return Bluebird.fromCallback(function(cb) {
     return self.closeBecause("Goodbye", defs.constants.REPLY_SUCCESS,
                     cb);
   });
+  // return new Promise(function(resolve, reject) {
```
Comment (kibertoad) [inline lib/channel_model.js:106]: Should this be commented out?

File: lib/channel_model.js [hunk 2]
```diff
@@ -74,10 +98,16 @@ C.open = function() {
 
 C.close = function() {
   var self = this;
-  return Promise.fromCallback(function(cb) {
+
+  return Bluebird.fromCallback(function(cb) {
```
Comment (kibertoad) [inline lib/channel_model.js:102]: Am I misunderstanding something, or promisify could be used here instead of Bluebird?

**Review by alexkvak — COMMENTED** [review]
> @alexkvak This needs to be synced with master after old Node versions were dropped

Done

File: lib/channel_model.js [hunk 3]
```diff
@@ -167,9 +197,18 @@ C.consume = function(queue, callback, options) {
   // NB we want the callback to be run synchronously, so that we've
   // registered the consumerTag before any messages can arrive.
   var fields = Args.consume(queue, options);
-  return Promise.fromCallback(function(cb) {
+  return Bluebird.fromCallback(function(cb) {
     self._rpc(defs.BasicConsume, fields, defs.BasicConsumeOk, cb);
   })
+  // return new Promise(function(resolve, reject) {
```
Comment (kibertoad) [inline lib/channel_model.js:203]: Should this be commented out?

File: lib/channel_model.js (hunk 2 above)
Comment (alexkvak) [reply lib/channel_model.js:102]: After dropping old Node version — it could

File: lib/channel_model.js (hunk 1 above)
Comment (alexkvak) [reply lib/channel_model.js:106]: I need some help with this.

If I replace `Bluebird.fromCallback` with commented code it seems that `channel.accept` is called _after_ this promise resolved. So the behavior changes and I have no idea why. I'm sure you have more context with this.

File: lib/channel_model.js (hunk 2 above)
Comment (kibertoad) [reply lib/channel_model.js:102]: Only Node 10+ is supported in master, so should be fine to use it now.

File: lib/callback_model.js [hunk 4]
```diff
@@ -5,7 +5,6 @@
 'use strict';
 
 var defs = require('./defs');
-var Promise = require('bluebird');
 var inherits = require('util').inherits;
```
Comment (jimmywarting) [inline lib/callback_model.js:8]: Unrelated to this PR but i just notices inherits, NodeJS docs mention discourage use of inherits and that class extends should be used instead.

Comment (kibertoad) [reply lib/callback_model.js:8]: I'll open PR for this

File: lib/channel_model.js (hunk 2 above)
Comment (alexkvak) [reply lib/channel_model.js:73]: Replaced with util.promisify

Comment (alexkvak) [reply lib/channel_model.js:73]: btw .travis.yml in main still builds library for old Node.js versions

Comment (kibertoad) [reply lib/channel_model.js:73]: Yes, I think it was just missed in https://github.com/squaremo/amqp.node/commit/267dd4e17c93efc755e40a4645510fdf70b9ebb8. Probably Travis needs to be adjusted to only test for Node 10+ as well.

**Comment by KristjanTammekivi** [comment]
Should also edit the types once this is ready https://github.com/DefinitelyTyped/DefinitelyTyped/tree/master/types/amqplib

File: channel_api.js [hunk 5]
```diff
@@ -1,15 +1,10 @@
-var raw_connect = require('./lib/connect').connect;
+const { promisify } = require('util');
+const raw_connect = promisify(require('./lib/connect').connect);
```
Comment (jimmywarting) [inline channel_api.js:2]: promisify is undefined

**Comment by xamgore** [comment]
I don't really recommend changing anything until tests are being migrated to jest and a proper [memory leak detector](https://www.npmjs.com/package/jest-leak-detector) is set up. There is a [PR](https://github.com/squaremo/amqp.node/pull/584) that fixes a memory leak, though either it isn't covered by tests or mocha's detector can't catch the case. Anyway, any change in the codebase may bring undesired behaviour in the future, and it may take time to fix and release it. As I may see in the comments here, Bluebird's semantics differs a bit from Node's promise implementation.

Also, @squaremo is quite a busy man and merges PRs if only 100% sure they are ok.

**Comment by carlhoerberg** [comment]
https://github.com/cloudamqp/amqp-client.js/ is a modern zero dependency AMQP library, implementing native promises, among other improvements.

**Comment by alexkvak** [comment]
> https://github.com/cloudamqp/amqp-client.js/ is a modern zero dependency AMQP library, implementing native promises, among other improvements.

Thanks a lot! I'll take a look and return with feedback 👍

**Comment by mohd-akram** [comment]
Use of native promises has been released in v0.10.0 via #689. This can be closed.

**Pull request closed without merging**